### PR TITLE
crypto: implement new key format

### PR DESCRIPTION
**Pull request by Glumli**

### CODE REVIEW

File: src/lib/crypto/index.js [hunk 1]
```diff
@@ -128,15 +125,64 @@ const getSPKIFromKey = key =>
...
+    case 'ck': // commonKey
+    case 'dk': // dataKey
+    case 'ak': // attachmentKey
+        return importRawKey(key.sym);
+    case 'privU': // userPrivateKey
+    case 'privA': // appPrivateKey
+        return getKeyFromPKCS8(key.priv);
+    case 'pubU': // userPublicKey
+    case 'pubA': // appPublicKey
+        return getKeyFromSPKI(key.pub);
```
Comment (s-urbaniak) [inline src/lib/crypto/index.js]: the above naming seems inconsistent (get vs import)

File: src/lib/crypto/index.js [hunk 2]
```diff
@@ -128,15 +125,64 @@ const getSPKIFromKey = key =>
...
+    default:
+        throw new Error('Not the right keyformat');
```
Comment (s-urbaniak) [inline src/lib/crypto/index.js]: `'invalid key format'`

File: src/lib/crypto/index.js [hunk 3]
```diff
@@ -226,31 +272,32 @@ const asymEncryptString = (publicKeySPKI, string) =>
...
 
 /**
  * Creates a random symmetric key.
  *
- * @returns {Promise} Resolves to a symmetric key as a jwk object
+ * @param {String} type of the key
+ * @returns {Promise} Resolves to a symmetric key as a hcKey object
  */
-const generateSymKey = () =>
+const generateSymKey = type =>
```
Comment (s-urbaniak) [inline src/lib/crypto/index.js]: unrelated, but didn't we switch to aes-gcm?

Comment (Glumli) [reply src/lib/crypto/index.js]: We decided to switch to aes gcm. Doesn't mean that it happened already ;)

Comment (s-urbaniak) [reply src/lib/crypto/index.js]: ohhhh ;-)

Comment (Glumli) [reply src/lib/crypto/index.js]: But I just created a Ticket

File: src/lib/crypto/index.js [hunk 4]
```diff
@@ -123,20 +120,69 @@ const getKeyFromSPKI = (SPKI) => {
...
+    case 'privU': // userPrivateKey
+    case 'privA': // appPrivateKey
+        return getPKCS8FromPrivateKey(key)
+            .then(keyString => ({
+                t: type,
+                v: 1,
+                priv: keyString,
+            }));
+    case 'pubU': // userPublicKey
+    case 'pubA': // appPublicKey
```
Comment (ibihim) [inline src/lib/crypto/index.js]: yeah, as mentioned above, you need some constant to handle this. if you change something in the naming of ck, you need to change it now at two places.

File: src/lib/crypto/index.js [hunk 5]
```diff
@@ -290,27 +337,30 @@ const generateAsymKeyPair = () =>
...
-            getPKCS8FromKey(keyPair.privateKey),
+            exportKey(keyPair.publicKey, `pub${type}`),
+            exportKey(keyPair.privateKey, `priv${type}`),
         ]))
-        .then(jwks => ({
-            publicKey: jwks[0],
-            privateKey: jwks[1],
+        .then(hcKeys => ({
+            publicKey: hcKeys[0],
+            privateKey: hcKeys[1],
```
Comment (ibihim) [inline src/lib/crypto/index.js]: this is not a change request:
what about 
```JavaScrip
.then(([publicKey, privateKey]) => ({ publicKey, privateKey }))
```

File: src/services/cryptoService.js [hunk 6]
```diff
@@ -15,7 +15,7 @@ const createEncryptData = encryptionMethod => commonKeyPromise =>
                 if (encryptedDataKey) {
                     return cryptoLib.symDecryptObject(commonKey, encryptedDataKey);
                 }
-                return cryptoLib.generateSymKey();
+                return cryptoLib.generateSymKey('dk');
```
Comment (ibihim) [inline src/services/cryptoService.js]: enum / const!

File: src/lib/crypto/index.js [hunk 7]
```diff
@@ -123,20 +120,69 @@ const getKeyFromSPKI = (SPKI) => {
...
+    switch (key.t) {
+    case 'ck': // commonKey
+    case 'dk': // dataKey
+    case 'ak': // attachmentKey
+        return getSymKeyFromString(key.sym);
+    case 'privU': // userPrivateKey
+    case 'privA': // appPrivateKey
+        return getPrivateKeyFromPKCS8(key.priv);
+    case 'pubU': // userPublicKey
+    case 'pubA': // appPublicKey
```
[the remark this answers is not in the document]
Comment (Glumli) [reply src/lib/crypto/index.js]: thought about it as well 👍

File: src/lib/crypto/index.js (hunk 5 above)
Comment (Glumli) [reply src/lib/crypto/index.js]: 👍

File: src/lib/crypto/index.js [hunk 8]
```diff
@@ -123,20 +120,69 @@ const getKeyFromSPKI = (SPKI) => {
...
+        return getPublicKeyFromSPKI(key.pub);
+    default:
+        throw new Error('invalid key type');
+    }
+};
+
+const exportKey = (key, type) => {
+    switch (type) {
+    case 'ck': // commonKey
```
Comment (Harsimran1) [inline src/lib/crypto/index.js]: we should mention `tek` here too probably if i get this right.

**Comment by Glumli** [comment]
@ibihim  PTAL

**Pull request merged**